File: python/src/openharness/builtin_tools.py
"""A few safe, dependency-free tools that the harness and CLI enable by default."""

from __future__ import annotations

import ast
import math
import operator
import re
from datetime import datetime, timezone
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError

import httpx

from .errors import ToolError
from .tools import tool

_OPS = {
    ast.Add: operator.add,
    ast.Sub: operator.sub,
    ast.Mult: operator.mul,
    ast.Div: operator.truediv,
    ast.FloorDiv: operator.floordiv,
    ast.Mod: operator.mod,
    ast.Pow: operator.pow,
    ast.USub: operator.neg,
    ast.UAdd: operator.pos,
}
_FUNCS = {
    name: getattr(math, name)
    for name in (
        "sqrt",
        "sin",
        "cos",
        "tan",
        "asin",
        "acos",
        "atan",
        "log",
        "log10",
        "log2",
        "exp",
        "floor",
        "ceil",
        "factorial",
        "radians",
        "degrees",
    )
}
_FUNCS.update({"abs": abs, "round": round, "min": min, "max": max})
_CONSTS = {"pi": math.pi, "e": math.e, "tau": math.tau}


def safe_eval(expression: str) -> float | int:
    """Evaluate arithmetic without ``eval``: numbers, + - * / // % **, and math functions."""

    def ev(node: ast.AST) -> float | int:
        if isinstance(node, ast.Expression):
            return ev(node.body)
        if isinstance(node, ast.Constant) and isinstance(node.value, (int, float)):
            return node.value
        if isinstance(node, ast.BinOp) and type(node.op) in _OPS:
            left, right = ev(node.left), ev(node.right)
            if isinstance(node.op, ast.Pow) and abs(right) > 1000:
                raise ToolError("Exponent too large")
            return _OPS[type(node.op)](left, right)
        if isinstance(node, ast.UnaryOp) and type(node.op) in _OPS:
            return _OPS[type(node.op)](ev(node.operand))
        if isinstance(node, ast.Name) and node.id in _CONSTS:
            return _CONSTS[node.id]
        if (
            isinstance(node, ast.Call)
            and isinstance(node.func, ast.Name)
            and node.func.id in _FUNCS
            and not node.keywords
        ):
            return _FUNCS[node.func.id](*[ev(a) for a in node.args])
        raise ToolError(f"Unsupported expression: {ast.dump(node)[:80]}")

    try:
        return ev(ast.parse(expression.replace("^", "**"), mode="eval"))
    except (SyntaxError, ZeroDivisionError, ValueError, OverflowError) as e:
        raise ToolError(f"Could not evaluate '{expression}': {e}") from e


@tool
def calculator(expression: str) -> str:
    """Evaluate a math expression exactly. Use this instead of doing arithmetic in your head.

    Args:
        expression: For example "(12.5 * 4) / 3" or "sqrt(2) ** 3". Supports + - * / // % ** and
            sqrt, log, sin, cos, tan, exp, floor, ceil, abs, round, min, max, pi, e.
    """
    result = safe_eval(expression)
    if isinstance(result, float) and result.is_integer() and abs(result) < 1e15:
        result = int(result)
    return str(result)


@tool
def current_time(timezone_name: str = "UTC") -> str:
    """Get the current date and time.

    Args:
        timezone_name: IANA time zone such as "America/Toronto" or "Europe/London".
    """
    try:
        tz = ZoneInfo(timezone_name) if timezone_name.upper() != "UTC" else timezone.utc
    except ZoneInfoNotFoundError as e:
        raise ToolError(f"Unknown time zone '{timezone_name}'") from e
    now = datetime.now(tz)
    return now.strftime("%A, %Y-%m-%d %H:%M:%S %Z (UTC%z)")


_TAGS = re.compile(r"<(script|style)[^>]*>.*?</\1>|<[^>]+>", re.DOTALL | re.IGNORECASE)


@tool(needs_approval=True)
async def fetch_url(url: str, max_chars: int = 8000) -> str:
    """Download a web page and return its text (HTML tags removed).

    Args:
        url: An http or https URL.
        max_chars: Maximum characters to return.
    """
    if not url.startswith(("http://", "https://")):
        raise ToolError("Only http and https URLs are allowed")
    async with httpx.AsyncClient(timeout=20, follow_redirects=True) as client:
        resp = await client.get(url, headers={"User-Agent": "openharness"})
    text = resp.text
    if "html" in resp.headers.get("content-type", ""):
        text = re.sub(r"\s+", " ", _TAGS.sub(" ", text))
    return f"HTTP {resp.status_code}\n{text.strip()[:max_chars]}"


DEFAULT_TOOLS = [calculator, current_time]
ALL_TOOLS = {"calculator": calculator, "current_time": current_time, "fetch_url": fetch_url}

File: typescript/src/builtinTools.ts
/** A few safe, dependency-free tools that the harness and CLI enable by default. */

import { ToolError } from "./errors.js";
import { s } from "./schema.js";
import { tool, type Tool } from "./tools.js";

const FUNCS: Record<string, (...a: number[]) => number> = {
  sqrt: Math.sqrt, sin: Math.sin, cos: Math.cos, tan: Math.tan, asin: Math.asin, acos: Math.acos, atan: Math.atan,
  log: Math.log, log10: Math.log10, log2: Math.log2, exp: Math.exp, floor: Math.floor, ceil: Math.ceil, abs: Math.abs,
  round: Math.round, min: Math.min, max: Math.max,
  radians: (d) => (d * Math.PI) / 180, degrees: (r) => (r * 180) / Math.PI,
  factorial: (n) => {
    if (!Number.isInteger(n) || n < 0 || n > 170) throw new ToolError("factorial needs an integer between 0 and 170");
    let r = 1;
    for (let i = 2; i <= n; i++) r *= i;
    return r;
  },
};
const CONSTS: Record<string, number> = { pi: Math.PI, e: Math.E, tau: 2 * Math.PI };

/** Evaluate arithmetic without `eval`: numbers, + - * / // % ** ^, parentheses and math functions. */
export function safeEval(expression: string): number {
  const tokens = expression.match(/\d+\.?\d*(?:e[+-]?\d+)?|\.\d+|\*\*|\/\/|[a-z_][a-z0-9_]*|[-+*/%^(),]|\S/gi) ?? [];
  let pos = 0;
  const peek = () => tokens[pos];
  const take = (t?: string) => {
    const tok = tokens[pos++];
    if (t !== undefined && tok !== t) throw new ToolError(`Could not evaluate '${expression}': expected '${t}'`);
    return tok;
  };
  const expr = (): number => {
    let v = term();
    while (peek() === "+" || peek() === "-") v = take() === "+" ? v + term() : v - term();
    return v;
  };
  const term = (): number => {
    let v = unary();
    for (;;) {
      const op = peek();
      if (op === "*") (take(), (v *= unary()));
      else if (op === "/") (take(), (v /= unary()));
      else if (op === "//") (take(), (v = Math.floor(v / unary())));
      else if (op === "%") (take(), (v %= unary()));
      else return v;
    }
  };
  const unary = (): number => {
    if (peek() === "-") return take(), -unary();
    if (peek() === "+") return take(), unary();
    return power();
  };
  const power = (): number => {
    const base = atom();
    if (peek() === "**" || peek() === "^") {
      take();
      const exp = unary();
      if (Math.abs(exp) > 1000) throw new ToolError("Exponent too large");
      return base ** exp;
    }
    return base;
  };
  const atom = (): number => {
    const t = take();
    if (t === undefined) throw new ToolError(`Could not evaluate '${expression}': unexpected end`);
    if (t === "(") {
      const v = expr();
      take(")");
      return v;
    }
    if (/^[\d.]/.test(t)) return Number(t);
    const name = t.toLowerCase();
    if (name in CONSTS) return CONSTS[name];
    if (name in FUNCS && peek() === "(") {
      take("(");
      const args: number[] = [];
      if (peek() !== ")") {
        args.push(expr());
        while (peek() === ",") (take(), args.push(expr()));
      }
      take(")");
      return FUNCS[name](...args);
    }
    throw new ToolError(`Unsupported expression near '${t}'`);
  };
  const value = expr();
  if (pos < tokens.length) throw new ToolError(`Could not evaluate '${expression}': unexpected '${tokens[pos]}'`);
  if (!Number.isFinite(value)) throw new ToolError(`Could not evaluate '${expression}': result is not finite`);
  return value;
}

export const calculator: Tool = tool({
  name: "calculator",
  description: "Evaluate a math expression exactly. Use this instead of doing arithmetic in your head.",
  parameters: s.object({
    expression: s
      .string()
      .describe('For example "(12.5 * 4) / 3" or "sqrt(2) ** 3". Supports + - * / // % ** and sqrt, log, sin, cos, tan, exp, floor, ceil, abs, round, min, max, pi, e.'),
  }),
  execute: ({ expression }) => {
    const v = safeEval(expression);
    return String(Number.isInteger(v) ? v : Number(v.toPrecision(15)));
  },
});

export const currentTime: Tool = tool({
  name: "current_time",
  description: "Get the current date and time.",
  parameters: s.object({ timezone_name: s.string().describe('IANA time zone such as "America/Toronto" or "Europe/London".').default("UTC") }),
  execute: ({ timezone_name }) => {
    const tz = timezone_name ?? "UTC";
    try {
      return new Intl.DateTimeFormat("en-CA", {
        timeZone: tz,
        weekday: "long",
        year: "numeric",
        month: "2-digit",
        day: "2-digit",
        hour: "2-digit",
        minute: "2-digit",
        second: "2-digit",
        hour12: false,
        timeZoneName: "short",
      }).format(new Date());
    } catch {
      throw new ToolError(`Unknown time zone '${tz}'`);
    }
  },
});

export const fetchUrl: Tool = tool({
  name: "fetch_url",
  description: "Download a web page and return its text (HTML tags removed).",
  parameters: s.object({
    url: s.string().describe("An http or https URL."),
    max_chars: s.integer().describe("Maximum characters to return.").default(8000),
  }),
  needsApproval: true,
  execute: async ({ url, max_chars }) => {
    if (!/^https?:\/\//.test(url)) throw new ToolError("Only http and https URLs are allowed");
    const resp = await fetch(url, { headers: { "User-Agent": "openharness" }, signal: AbortSignal.timeout(20000) });
    let text = await resp.text();
    if ((resp.headers.get("content-type") ?? "").includes("html"))
      text = text.replace(/<(script|style)[^>]*>[\s\S]*?<\/\1>|<[^>]+>/gi, " ").replace(/\s+/g, " ");
    return `HTTP ${resp.status}\n${text.trim().slice(0, max_chars ?? 8000)}`;
  },
});

export const DEFAULT_TOOLS: Tool[] = [calculator, currentTime];
export const ALL_TOOLS: Record<string, Tool> = { calculator, current_time: currentTime, fetch_url: fetchUrl };
